### PR TITLE
[BCEL-321] AbstractClassPathRepository to share findClass logic for repositories

### PR DESCRIPTION
**Pull request by suztomo**
# Problem to solve
BCEL has different Repository classes with slight modification in its underlying cache:

- ClassPathRepository uses HashMap<String, JavaClass> for JavaClass cache
- MemorySensitiveClassPathRepository uses HashMap<String, SoftReference<JavaClass>
- New LruCacheClassPathRepository by BCEL-320 will use LinkedHashMap<String, JavaClass> for JavaClass cache

The logic of loadClass, storeClass, and findClass methods are almost same ([attached screenshot in BCEL-321](https://issues.apache.org/jira/browse/BCEL-321)). I think they can leverage an abstraction over the internal cache so that they will have less duplicate code.

# Solution

This PR introduces AbstractClassPathRepository to share findClass logic for these repositories. The subclasses focus on how to cache the result via storeClass and findClass methods.

# jacoco report

## Coverage

image

## No incompatible changes
image

### CODE REVIEW

File: src/main/java/org/apache/bcel/util/ClassPathRepository.java [hunk 1]
```diff
@@ -65,95 +61,6 @@ public JavaClass findClass(final String className) {
         return _loadedClasses.get(className);
     }
 
-    /**
```
Comment (suztomo) [inline src/main/java/org/apache/bcel/util/ClassPathRepository.java]: The findClass logic has moved to AbstractClassPathRepository.

File: src/main/java/org/apache/bcel/util/MemorySensitiveClassPathRepository.java [hunk 2]
```diff
@@ -71,96 +68,6 @@ public JavaClass findClass(final String className) {
         return ref.get();
     }
 
-    /**
```
Comment (suztomo) [inline src/main/java/org/apache/bcel/util/MemorySensitiveClassPathRepository.java]: The findClass logic has moved to AbstractClassPathRepository.

File: src/main/java/org/apache/bcel/util/AbstractClassPathRepository.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * This abstract class provides a logic of a loading {@link JavaClass} objects class names via {@link ClassPath}.
+ *
+ * <p>Subclasses can choose caching strategy of the objects by implementing the abstract methods (e.g., {@link
+ * #storeClass(JavaClass)} and {@link #findClass(String)}).
+ *
```
Comment (garydgregory) [inline src/main/java/org/apache/bcel/util/AbstractClassPathRepository.java]: Missing close tag </p>

File: src/main/java/org/apache/bcel/util/AbstractClassPathRepository.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+     *
+     * @param className
+     *            the name of the class
+     * @return the JavaClass object
+     * @throws ClassNotFoundException
+     *             if the class is not in the Repository, and could not be found on the classpath
+     */
+    @Override
+    public JavaClass loadClass(String className) throws ClassNotFoundException {
+        if ((className == null) || className.isEmpty()) {
```
Comment (garydgregory) [inline src/main/java/org/apache/bcel/util/AbstractClassPathRepository.java]: Don't need parens in `(className == null)`

File: src/test/java/org/apache/bcel/util/ClassPathRepositoryTestCase.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import org.junit.Test;
+
+/**
+ * Tests {@link ClassPathRepository} and {@link MemorySensitiveClassPathRepository}.
+ *
+ * <p>Without memory scarcity, these two classes behave in the same manner.
+ */
+public class ClassPathRepositoryTestCase {
+
+    private void verifyCaching(AbstractClassPathRepository repository) throws ClassNotFoundException {
```
Comment (garydgregory) [inline src/test/java/org/apache/bcel/util/ClassPathRepositoryTestCase.java]: Please run `mvn -P jacoco -P japicmp install site' and make sure you get as close to 100% code coverage in the reports for the new and updated code. Also make sure the JApiCmp report does not report BC breakage.

File: src/main/java/org/apache/bcel/util/AbstractClassPathRepository.java (hunk 3 above)
Comment (suztomo) [reply src/main/java/org/apache/bcel/util/AbstractClassPathRepository.java]: Added close tag.

File: src/main/java/org/apache/bcel/util/AbstractClassPathRepository.java (hunk 4 above)
Comment (suztomo) [reply src/main/java/org/apache/bcel/util/AbstractClassPathRepository.java]: Updated.

File: src/test/java/org/apache/bcel/util/ClassPathRepositoryTestCase.java (hunk 5 above)
Comment (suztomo) [reply src/test/java/org/apache/bcel/util/ClassPathRepositoryTestCase.java]: `mvn site` is failing. Created #30 .

Comment (suztomo) [reply src/test/java/org/apache/bcel/util/ClassPathRepositoryTestCase.java]: Added more tests to increase test coverage. Added coverage information in the description of this PR.

I found corner cases with IOException were hard to cover.

image
